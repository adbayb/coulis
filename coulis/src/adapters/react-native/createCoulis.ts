import type { CreateCoulis } from "../../core/ports/createCoulis";
import { compose, isObject } from "../../core/entities/primitive";
import { transformDimension } from "./transformers";
import { createUnsupportedLogger } from "./helpers";

type CreateCoulisOutput = Record<string, unknown>;

export const createCoulis: CreateCoulis<{
	Input: { WithCSSVariables: false };
	Output: CreateCoulisOutput;
}> = (contract) => {
	const unsupportedLogger = createUnsupportedLogger();

	const properties = contract.properties(
		contract.theme as Parameters<typeof contract.properties>[0],
	);

	const shorthands = (contract.shorthands ?? {}) as NonNullable<
		typeof contract.shorthands
	>;

	const shorthandNames = Object.keys(shorthands);

	const isCustomShorthandProperty = (name: string) => {
		return shorthandNames.includes(name);
	};

	const getStyleValue = (name: string, value: unknown) => {
		const getValue = (styleValue: unknown): unknown => {
			const transform = compose(transformDimension);
			const propertyValue = properties[name as keyof typeof properties];

			return transform({
				name,
				value:
					typeof propertyValue === "function"
						? // eslint-disable-next-line @typescript-eslint/no-unsafe-call
							propertyValue(styleValue)
						: isObject(propertyValue)
							? propertyValue[styleValue as string]
							: styleValue,
			}).value;
		};

		if (isObject(value)) {
			unsupportedLogger.behavior(
				"States are not supported, ignoring non-base values.",
			);

			return getValue(value.base);
		}

		return getValue(value);
	};

	return {
		createKeyframes() {
			unsupportedLogger.method("createKeyframes");

			return {};
		},
		createStyles(input) {
			// TODO: cache (insert method)
			const styles: CreateCoulisOutput = {};

			for (const propertyName of Object.keys(input)) {
				const value = input[propertyName as keyof typeof input];

				if (isCustomShorthandProperty(propertyName)) {
					const shorthandedPropertyNames = shorthands[
						propertyName
					] as string[] | undefined;

					if (shorthandedPropertyNames === undefined) continue;

					for (const shorthandedPropertyName of shorthandedPropertyNames) {
						styles[shorthandedPropertyName] = getStyleValue(
							shorthandedPropertyName,
							value,
						);
					}
				} else {
					styles[propertyName] = getStyleValue(propertyName, value);
				}
			}

			console.log(styles);

			return styles;
		},
		getContract() {
			return {
				propertyNames: [
					...shorthandNames,
					...Object.keys(properties),
				] as ReturnType<typeof this.getContract>["propertyNames"],
			};
		},
		getMetadata() {
			unsupportedLogger.method("getMetadata");

			return [];
		},
		setGlobalStyles() {
			unsupportedLogger.method("setGlobalStyles");
		},
	};
};
