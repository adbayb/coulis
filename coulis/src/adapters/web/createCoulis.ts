import type { CreateCoulis } from "../../core/ports/createCoulis";
import { STYLE_TYPES } from "../../core/entities/style";
import type { StyleType } from "../../core/entities/style";
import { isNumber, isObject } from "../../core/entities/primitive";
import type { RecordLike } from "../../core/entities/primitive";
import { createMapCache, createSetCache } from "../../core/entities/cache";
import type { SetCache } from "../../core/entities/cache";
import type { ClassName, Rule, StyleSheet } from "./types";
import { createVirtualStyleSheet } from "./stylesheet/virtual";
import { createDomStyleSheet } from "./stylesheet/dom";
import {
	createClassName,
	createCustomProperties,
	createDeclaration,
	getEvaluatedTemplate,
	isShorthandProperty,
} from "./helpers";
import { IS_SERVER_ENVIRONMENT } from "./constants";

export const createCoulis: CreateCoulis<{
	Input: {
		WithCSSVariables: true;
	};
	Output: ClassName;
	// eslint-disable-next-line sonarjs/max-lines-per-function
}> = (contract) => {
	const createStyleSheet = IS_SERVER_ENVIRONMENT
		? createVirtualStyleSheet
		: createDomStyleSheet;

	const styleSheetByTypeAdaptee = STYLE_TYPES.reduce(
		(output, type) => {
			output[type] = createStyleSheet(type);

			return output;
		},
		{} as Record<StyleType, StyleSheet>,
	);

	const classNameByTypeCache = createMapCache<
		StyleType,
		SetCache<ClassName>
	>();

	const hydratedClassNameCache = new Set(
		Object.values(styleSheetByTypeAdaptee).flatMap((styleSheet) =>
			styleSheet.getHydratedClassNames(),
		),
	);

	const shorthands = (contract.shorthands ?? {}) as NonNullable<
		typeof contract.shorthands
	>;

	const shorthandNames = Object.keys(shorthands);
	let collectedCustomProperties = "";

	const properties = contract.properties(
		(contract.theme &&
			createCustomProperties(contract.theme, (name, value) => {
				collectedCustomProperties += `${name}:${String(value)};`;
			})) as Parameters<typeof contract.properties>[0],
	);

	const isCustomShorthandProperty = (name: string) => {
		return shorthandNames.includes(name);
	};

	const getDeclaration = ({
		name,
		value,
	}: {
		name: keyof RecordLike;
		value: RecordLike[keyof RecordLike];
	}) => {
		const propertyValue = properties[name as keyof typeof properties];

		return createDeclaration({
			name,
			value:
				typeof propertyValue === "function"
					? // eslint-disable-next-line @typescript-eslint/no-unsafe-call
						propertyValue(value)
					: isObject(propertyValue)
						? (propertyValue[value as string] ?? value)
						: value,
		});
	};

	const createDeclarationBlock = (input: RecordLike) => {
		let declarationBlock = "";
		const propertyNames = Object.keys(input);

		propertyNames.forEach((propertyName) => {
			const value = input[propertyName];

			if (value === undefined) return;

			if (isCustomShorthandProperty(propertyName)) {
				const shorthandedPropertyNames = shorthands[propertyName];

				if (shorthandedPropertyNames === undefined) return;

				shorthandedPropertyNames.forEach((shorthandedPropertyName) => {
					declarationBlock += getDeclaration({
						name: shorthandedPropertyName as string,
						value,
					});
				});
			} else {
				declarationBlock += getDeclaration({
					name: propertyName,
					value,
				});
			}
		});

		return declarationBlock;
	};

	const insert = ({
		cacheInput,
		onCreateRule,
		type,
	}: {
		cacheInput: string;
		onCreateRule: (input: { className: ClassName }) => Rule;
		type: StyleType;
	}): ClassName => {
		const className = createClassName(cacheInput);

		if (hydratedClassNameCache.has(className)) return className;

		let cache = classNameByTypeCache.get(type);

		if (!cache) {
			cache = createSetCache();
			classNameByTypeCache.add(type, cache);
		}

		if (cache.has(className)) return className;

		cache.add(className);
		styleSheetByTypeAdaptee[type].insert(
			className,
			onCreateRule({ className }),
		);

		return className;
	};

	insert({
		cacheInput: collectedCustomProperties,
		onCreateRule() {
			return `:root{${collectedCustomProperties}}`;
		},
		type: "global",
	});

	return {
		createKeyframes(input) {
			return insert({
				cacheInput: JSON.stringify(input),
				onCreateRule({ className }) {
					let rule = "";

					const selectors = Object.keys(
						input,
					) as (keyof typeof input)[];

					for (const selector of selectors) {
						const style = input[selector];

						if (!style) continue;

						const ruleSelector = isNumber(selector)
							? `${selector}%`
							: String(selector);

						rule += `${ruleSelector}{${createDeclarationBlock(style)}}`;
					}

					return `@keyframes ${className}{${rule}}`;
				},
				type: "global",
			});
		},
		createStyles(input) {
			const classNames: ClassName[] = [];

			const collectClassNames = (name: string, value: unknown) => {
				const isShorthandPropertyValue = isShorthandProperty(name);

				let type: StyleType = isShorthandPropertyValue
					? "shorthand"
					: "longhand";

				if (!isObject(value)) {
					const declaration = getDeclaration({
						name,
						value,
					});

					classNames.push(
						insert({
							cacheInput: declaration,
							onCreateRule({ className }) {
								return `.${className}{${declaration}}`;
							},
							type,
						}),
					);

					return;
				}

				const stateKeys = Object.keys(value);

				for (const stateKey of stateKeys) {
					const stateValue = value[stateKey];

					const declaration = getDeclaration({
						name,
						value: stateValue,
					});

					const isBaseState = stateKey === "base";

					const stateTemplate = isBaseState
						? "coulis[selector]{coulis[declaration]}"
						: contract.states?.[stateKey];

					if (stateTemplate === undefined) continue;

					const preComputedRule = getEvaluatedTemplate(
						stateTemplate,
						{
							declaration,
							selector: `.coulis[className]`,
						},
					);

					if (preComputedRule.startsWith("@")) {
						type = isShorthandPropertyValue
							? "atShorthand"
							: "atLonghand";
					}

					classNames.push(
						insert({
							/*
							 * The key is not included to compute the className when `key` equals to "base" as base is equivalent to an unconditional value.
							 * This exclusion will allow to recycle cache if the style value has been already defined unconditionally.
							 */
							cacheInput: isBaseState
								? declaration
								: `${stateKey}${declaration}`,
							onCreateRule({ className }) {
								return preComputedRule.replaceAll(
									"coulis[className]",
									className,
								);
							},
							type,
						}),
					);
				}
			};

			for (const propertyName of Object.keys(input)) {
				const value = input[propertyName as keyof typeof input];

				if (isCustomShorthandProperty(propertyName)) {
					const shorthandedPropertyNames = shorthands[propertyName];

					if (shorthandedPropertyNames === undefined) continue;

					for (const shorthandedPropertyName of shorthandedPropertyNames) {
						collectClassNames(
							shorthandedPropertyName as string,
							value,
						);
					}
				} else {
					collectClassNames(propertyName, value);
				}
			}

			return classNames.join(" ");
		},
		getContract() {
			return {
				propertyNames: [
					...shorthandNames,
					...Object.keys(properties),
				] as ReturnType<typeof this.getContract>["propertyNames"],
			};
		},
		getMetadata() {
			const metadata = STYLE_TYPES.map((type) => {
				const { getContent } = styleSheetByTypeAdaptee[type];
				const cachedClassNames = classNameByTypeCache.get(type);
				const content = getContent();

				return {
					attributes: {
						"data-coulis-cache": [
							...(cachedClassNames?.getAll() ?? []),
						].join(","),
						"data-coulis-type": type,
					},
					content,
				};
			});

			metadata.toString = () => {
				return metadata.reduce((output, { attributes, content }) => {
					const stringifiedAttributes = (
						Object.keys(attributes) as (keyof typeof attributes)[]
					)
						.map(
							// eslint-disable-next-line sonarjs/no-nested-functions
							(attributeKey) =>
								`${attributeKey}="${attributes[attributeKey]}"`,
						)
						.join(" ");

					output += `<style ${stringifiedAttributes}>${content}</style>`;

					return output;
				}, "");
			};

			return metadata;
		},
		setGlobalStyles(input) {
			insert({
				cacheInput: JSON.stringify(input),
				onCreateRule() {
					let rule = "";
					const selectors = Object.keys(input);

					for (const selector of selectors) {
						const style = input[selector];

						if (style === undefined) continue; // TODO: fix coulis side undefined value (see bienvenuemarket)

						rule +=
							typeof style === "string"
								? `${selector} ${style};`
								: `${selector}{${createDeclarationBlock(style)}}`;
					}

					return rule;
				},
				type: "global",
			});
		},
	};
};
