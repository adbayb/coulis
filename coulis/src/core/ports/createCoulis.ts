import type { ThemeLike } from "../entities/theme";
import type { GlobalStyles, Styles } from "../entities/style";
import type { StatesLike } from "../entities/state";
import type { ShortandsLike } from "../entities/shorthand";
import type { PropertiesLike } from "../entities/property";
import type { RecordLike, WithNewLeafNodes } from "../entities/primitive";
import type { Keyframes } from "../entities/keyframe";

export type CreateCoulis<Output> = <
	const Properties extends PropertiesLike, // `const` needed to not widen strict type (for example, to avoid `width: [50, 100]` being inferred as `width: number`).
	Shorthands extends ShortandsLike<Properties> | undefined = undefined,
	States extends StatesLike | undefined = undefined,
	Theme extends ThemeLike | undefined = undefined,
>(contract: {
	properties: (
		theme: Theme extends RecordLike
			? WithNewLeafNodes<Theme, string>
			: undefined,
	) => Properties;
	shorthands?: Shorthands;
	states?: States;
	theme?: Theme;
}) => {
	createKeyframes: (input: Keyframes<Properties>) => Output;
	createStyles: (input: Styles<Properties, Shorthands, States>) => Output;
	getMetadata: () => {
		toString: () => string;
		value: {
			attributes: Record<
				"data-coulis-cache" | "data-coulis-type",
				string
			>;
			content: string;
		}[];
	};
	setGlobalStyles: (input: GlobalStyles<Properties>) => void;
};

const createCoulisFake: CreateCoulis<string> = (_input) => {
	return {
		createKeyframes() {
			return "fake";
		},
		createStyles() {
			return "fake";
		},
		getMetadata() {
			return {
				toString() {
					return "";
				},
				value: [],
			};
		},
		setGlobalStyles() {
			// No op
		},
	};
};

const { createStyles } = createCoulisFake({
	properties() {
		return {
			backgroundColor: {
				blue: "blue",
			},
		};
	},
	shorthands: {
		backgroundColorAlias: ["backgroundColor"],
	},
	states: {
		hover: "coulis[selector]:hover{coulis[declaration]}",
	},
});

createStyles({
	backgroundColor: {
		base: "blue",
	},
	backgroundColorAlias: "blue",
});
