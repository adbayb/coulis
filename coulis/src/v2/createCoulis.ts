import type { RecordLike } from "./types";

export type Coulis = {
	createIntermediateRepresentation: <Input extends RecordLike>(
		type: Type,
		input: Input,
	) => IntermediateRepresentation<Input>;
	getCache: () => Cache;
};

export type CreateCoulis = (createId: (input: RecordLike) => string) => Coulis;

export const createCoulis: CreateCoulis = (createId) => {
	const cache = TYPES.reduce((output, type) => {
		output[type] = new Map<
			IntermediateRepresentationId,
			IntermediateRepresentation<RecordLike>
		>();

		return output;
	}, {} as Cache);

	return {
		createIntermediateRepresentation<Input extends RecordLike>(
			type: Type,
			input: Input,
		) {
			const id = createId(input);
			const styles = cache[type].get(id);

			const output: IntermediateRepresentation<Input> = {
				id,
				isCached: false,
				payload: input,
			};

			if (!styles) {
				cache[type].set(id, output);
			}

			output.isCached = true;

			return output;
		},
		getCache() {
			return cache;
		},
	};
};

/**
 * The order is important to enforce the more precise properties take precedence over less precise ones.
 * Global properties has a lesser specificity than (<) shorthand ones:
 * global (e.g div { background-color }) < shorthand (e.g background) < longhand (e.g background-color) < conditional-shorthand (e.g @media { background }) < conditional-longhand (e.g @media { background-color }) properties.
 */
export const TYPES = Object.freeze([
	"global",
	"shorthand",
	"longhand",
	"atShorthand",
	"atLonghand",
] as const);

type IntermediateRepresentationId = string;

type IntermediateRepresentation<Payload extends RecordLike> = {
	id: IntermediateRepresentationId;
	isCached: boolean;
	payload: Payload;
};

export type Type = (typeof TYPES)[number];

type Cache = Record<
	Type,
	Map<IntermediateRepresentationId, IntermediateRepresentation<RecordLike>>
>;
