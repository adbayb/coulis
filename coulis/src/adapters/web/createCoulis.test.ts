import { describe, expect, test } from "vitest";

import { createCoulis } from "./createCoulis";

describe("createCoulis (web adapter)", () => {
	test("should create keyframes", () => {
		expect(
			createKeyframes({
				25: {
					transform: "scale(1.25)",
				},
				"50%": {
					transform: "scale(1.5)",
				},
				"from": {
					transform: "scale(1)",
				},
				"to": {
					transform: "scale(1)",
				},
			}),
		).toBe("c77e20e50");
	});

	test("should create styles", () => {
		expect(
			createStyles({
				backgroundColor: {
					alt: "surfacePrimary",
					base: "surfaceSecondary",
					hover: "neutralTransparent",
				},
				color: "neutralDark",
				padding: {
					base: 0,
				},
			}),
		).toBe("ce5eab527 ca4eaf829 ce899257c cdf15bb1c c8cf22ec5");
	});

	test("should extract styles", () => {
		expect(getMetadata()).toMatchSnapshot();
	});

	test("should extract styles given stringified styles", () => {
		expect(String(getMetadata())).toMatchSnapshot();
	});

	test("should get contract", () => {
		expect(getContract()).toStrictEqual({
			propertyNames: [
				"size",
				"backgroundColor",
				"boxSizing",
				"color",
				"colorScheme",
				"fontFamily",
				"height",
				"margin",
				"padding",
				"src",
				"transform",
				"width",
			],
		});
	});
});

const tokens = Object.freeze({
	colors: {
		black: "black",
		blue: [
			"rgb(241,244,248)",
			"rgb(226,232,240)",
			"rgb(201,212,227)",
			"rgb(168,186,211)",
			"rgb(119,146,185)",
		],
		transparent: "transparent",
		white: "white",
	},
	spacings: {
		0: "0px",
		0.5: "0.5rem",
		1: "1rem",
		1.5: "2rem",
	},
} as const);

const {
	createKeyframes,
	createStyles,
	getContract,
	getMetadata,
	setGlobalStyles,
} = createCoulis({
	properties(theme) {
		return {
			backgroundColor: theme.colors,
			boxSizing: true,
			color: theme.colors,
			colorScheme(input: "black" | "white") {
				return input === "black" ? "dark" : "light";
			},
			fontFamily: true,
			height: true,
			margin: theme.spacings,
			padding: theme.spacings,
			src: true,
			transform: true,
			width: [50, 100],
		};
	},
	shorthands: {
		size: ["width"],
	},
	states: {
		alt: "coulis[selector][alt]{coulis[declaration]}",
		hover: "coulis[selector]{coulis[declaration]}",
	},
	theme: {
		colors: {
			neutralDark: tokens.colors.black,
			neutralLight: tokens.colors.white,
			neutralTransparent: tokens.colors.transparent,
			surfacePrimary: tokens.colors.blue[4],
			surfaceSecondary: tokens.colors.blue[2],
		},
		spacings: tokens.spacings,
	},
});

// eslint-disable-next-line vitest/require-hook
setGlobalStyles({
	"*,*::before,*::after": {
		boxSizing: "inherit",
	},
	".globalClass+.otherGlobalClass": {
		colorScheme: "black",
	},
	"@charset": '"utf-8"',
	"@font-face": {
		fontFamily: "'AliasedHelvetica'",
		src: "local(Helvetica)",
	},
	"@import":
		"url('https://fonts.googleapis.com/css?family=Open+Sans&display=swap')",
	"html": {
		boxSizing: "border-box",
	},
	"html,body": {
		backgroundColor: "surfacePrimary",
		fontFamily: "Open Sans, AliasedHelvetica",
		margin: 0,
		padding: 0,
	},
});
