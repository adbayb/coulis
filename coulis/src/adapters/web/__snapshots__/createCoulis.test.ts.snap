// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createCoulis (web adapter) > should extract styles 1`] = `
{
  "toString": [Function],
  "value": [
    {
      "attributes": {
        "data-coulis-cache": "c66501570,cd2b4a084,c77e20e50",
        "data-coulis-type": "global",
      },
      "content": ":root{--colors-neutralDark:black;--colors-neutralLight:white;--colors-neutralTransparent:transparent;--colors-surfacePrimary:rgb(119,146,185);--colors-surfaceSecondary:rgb(201,212,227);--spacings-0:0px;--spacings-1:1rem;--spacings-0-5:0.5rem;--spacings-1-5:2rem;}*,*::before,*::after{box-sizing:inherit;}.globalClass+.otherGlobalClass{color-scheme:dark;}@charset "utf-8";@font-face{font-family:'AliasedHelvetica';src:local(Helvetica);}@import url('https://fonts.googleapis.com/css?family=Open+Sans&display=swap');html{box-sizing:border-box;}html,body{background-color:var(--colors-surfacePrimary);font-family:Open Sans, AliasedHelvetica;margin:var(--spacings-0);padding:var(--spacings-0);}@keyframes c77e20e50{25%{transform:scale(1.25);}50%{transform:scale(1.5);}from{transform:scale(1);}to{transform:scale(1);}}",
    },
    {
      "attributes": {
        "data-coulis-cache": "c8cf22ec5",
        "data-coulis-type": "shorthand",
      },
      "content": ".c8cf22ec5{padding:var(--spacings-0);}",
    },
    {
      "attributes": {
        "data-coulis-cache": "ce5eab527,ca4eaf829,ce899257c,cdf15bb1c",
        "data-coulis-type": "longhand",
      },
      "content": ".ce5eab527[alt]{background-color:var(--colors-surfacePrimary);}.ca4eaf829{background-color:var(--colors-surfaceSecondary);}.ce899257c{background-color:var(--colors-neutralTransparent);}.cdf15bb1c{color:var(--colors-neutralDark);}",
    },
    {
      "attributes": {
        "data-coulis-cache": "",
        "data-coulis-type": "atShorthand",
      },
      "content": "",
    },
    {
      "attributes": {
        "data-coulis-cache": "",
        "data-coulis-type": "atLonghand",
      },
      "content": "",
    },
  ],
}
`;

exports[`createCoulis (web adapter) > should extract styles given stringified styles 1`] = `"<style data-coulis-cache="c66501570,cd2b4a084,c77e20e50" data-coulis-type="global">:root{--colors-neutralDark:black;--colors-neutralLight:white;--colors-neutralTransparent:transparent;--colors-surfacePrimary:rgb(119,146,185);--colors-surfaceSecondary:rgb(201,212,227);--spacings-0:0px;--spacings-1:1rem;--spacings-0-5:0.5rem;--spacings-1-5:2rem;}*,*::before,*::after{box-sizing:inherit;}.globalClass+.otherGlobalClass{color-scheme:dark;}@charset "utf-8";@font-face{font-family:'AliasedHelvetica';src:local(Helvetica);}@import url('https://fonts.googleapis.com/css?family=Open+Sans&display=swap');html{box-sizing:border-box;}html,body{background-color:var(--colors-surfacePrimary);font-family:Open Sans, AliasedHelvetica;margin:var(--spacings-0);padding:var(--spacings-0);}@keyframes c77e20e50{25%{transform:scale(1.25);}50%{transform:scale(1.5);}from{transform:scale(1);}to{transform:scale(1);}}</style><style data-coulis-cache="c8cf22ec5" data-coulis-type="shorthand">.c8cf22ec5{padding:var(--spacings-0);}</style><style data-coulis-cache="ce5eab527,ca4eaf829,ce899257c,cdf15bb1c" data-coulis-type="longhand">.ce5eab527[alt]{background-color:var(--colors-surfacePrimary);}.ca4eaf829{background-color:var(--colors-surfaceSecondary);}.ce899257c{background-color:var(--colors-neutralTransparent);}.cdf15bb1c{color:var(--colors-neutralDark);}</style><style data-coulis-cache="" data-coulis-type="atShorthand"></style><style data-coulis-cache="" data-coulis-type="atLonghand"></style>"`;
